### PR TITLE
Fix 792

### PR DESCRIPTION
**Pull request by yzhang-gh**
On behalf of @Lemmingh.

### CODE REVIEW

**Comment by yzhang-gh** [comment]
Will do it tonight.

File: src/util.ts [hunk 1]
```diff
@@ -100,80 +101,90 @@ export function showChangelog() {
...
- * @param text A Markdown heading
+ * @param text - A Markdown heading content.
  */
-function mdHeadingToPlaintext(text: string) {
+function mdHeadingToPlaintext(text: string): string {
     //// Issue #515
     text = text.replace(/\[([^\]]*)\]\[[^\]]*\]/, (_, g1) => g1);
-    //// Escape leading `1.` and `1)` (#567, #585)
-    text = text.replace(/^([\d]+)(\.)/, (_, g1) => g1 + '%dot%');
-    text = text.replace(/^([\d]+)(\))/, (_, g1) => g1 + '%par%');
```
Comment (yzhang-gh) [inline src/util.ts:123]: Did you check this? It is kind of special as they are CommonMark syntaxes but...

It occurred to me whether we can properly deal with a heading like `# - pretends to be a list`.

**Review by yzhang-gh — COMMENTED** [review]
👍

File: package-lock.json [hunk 2]
```diff
@@ -507,6 +507,7 @@
             "integrity": "sha512-1Yj8h9Q+QDF5FzhMs/c9+6UntbD5MkRfRwac8DoEm9ZfUBZ7tZ55YcGVAzEe4bXsdQHEk+s9S5wsOKVdZrw0tQ==",
             "dev": true,
             "optional": true
+            "dev": true
```
Comment (Lemmingh) [inline package-lock.json]: @yzhang-gh,
Seems that you broke `package-lock.json`.
It's best regenerate it by npm.

File: package.json [hunk 3]
```diff
@@ -450,10 +450,10 @@
     },
     "devDependencies": {
         "@types/highlight.js": "^9.12.4",
-        "@types/image-size": "^0.7.0",
+        "@types/image-size": "^0.8.0",
         "@types/markdown-it": "^10.0.2",
         "@types/mocha": "^5.2.6",
-        "@types/node": "^12.12.58",
+        "@types/node": "^14.10.1",
```
Comment (Lemmingh) [inline package.json]: * You told me that [new `image-size` breaks the compatibility](https://github.com/yzhang-gh/vscode-markdown/commit/9b7fb3a5aa672847b288c6f87822dafb5b14ad6b#commitcomment-42278369).
* Due to **webpack** and **VS Code**, we have to stick to **Node.js 12** in the foreseeable future.

Thus, updating these two type definitions looks not a good idea.

**Review by Lemmingh — CHANGES_REQUESTED** [review]
Honestly, I always manipulate dependencies by `npm`, and update with extreme caution.

File: src/util.ts (hunk 1 above)
Comment (Lemmingh) [reply src/util.ts:123]: They are covered by `renderInline`.

File: package-lock.json (hunk 2 above)
Comment (yzhang-gh) [reply package-lock.json]: It's already regenerated (05f8d4f).

File: src/markdownEngine.ts [hunk 4]
```diff
@@ -111,7 +130,7 @@ class MarkdownEngine {
             if (originalHeadingOpen) {
                 return originalHeadingOpen(tokens, idx, options, env, self);
             } else {
-                return self.renderToken(tokens, idx, options, env, self);
+                return self.renderToken(tokens, idx, options);
```
[the remark this answers is not in the document]
Comment (Lemmingh) [reply src/markdownEngine.ts:133]: https://markdown-it.github.io/markdown-it/#Renderer.renderToken

File: package.json (hunk 3 above)
Comment (yzhang-gh) [reply package.json]: `image-size@0.8.0` looks fine (`0.9.1` doesn't).

> * Due to **webpack** and **VS Code**, we have to stick to **Node.js 12** in the foreseeable future.

I agree.

File: src/markdownEngine.ts [hunk 5]
```diff
@@ -2,13 +2,32 @@
...
 export const extensionBlacklist = new Set<string>(["vscode.markdown-language-features", "yzhang.markdown-all-in-one"]);
 
+interface IMarkdownEngine {
+}
+
+/**
+ * A strict CommonMark only engine powered by `markdown-it`.
+ */
+class CommonmarkEngine implements IMarkdownEngine {
+    readonly #engine: MarkdownIt;
```
[the remark this answers is not in the document]
Comment (Lemmingh) [reply src/markdownEngine.ts:18]: ECMAScript private field.

I think a Markdown engine instance, as defined in `interface IMarkdownEngine`, should expose a readonly `engine` property. So I wrote `public get engine()` below.

However, the instance needs to store the engine somewhere inside. So I created private field `#engine`.

[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:18]: 👌

File: src/markdownEngine.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:133]: 👌

File: src/util.ts (hunk 1 above)
Comment (yzhang-gh) [reply src/util.ts:123]: Nice!

File: src/markdownEngine.ts [hunk 6]
```diff
@@ -2,13 +2,32 @@
 
 import { extensions, workspace, WorkspaceConfiguration } from 'vscode';
 import { slugify } from './util';
-import { MarkdownContribution, MarkdownContributionProvider, getMarkdownContributionProvider } from './markdownExtensions'
+import { MarkdownContribution, MarkdownContributionProvider, getMarkdownContributionProvider } from './markdownExtensions';
+import MarkdownIt = require('markdown-it');
```
[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:6]: We have another (lazy) `import` statement here

https://github.com/yzhang-gh/vscode-markdown/blob/faa18bddd35618a99b73cc093919366eff9ea432/src/markdownEngine.ts#L49

although I don't think it improves performance much. Many features rely on it. Probably we also change this?

[the remark this answers is not in the document]
Comment (Lemmingh) [reply src/markdownEngine.ts:6]: To ... what? Could you please explain?

I don't know whether webpack will handle this case in some special way.

Also see my note:
https://github.com/users/Lemmingh/projects/2#card-44982812

[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:6]: (I mean `MarkdownIt` is already imported so we don't need to import it again.)

[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:6]: > To ... what? Could you please explain?

The activation time

[the remark this answers is not in the document]
Comment (Lemmingh) [reply src/markdownEngine.ts:6]: I see.

I agree to change `(await import('markdown-it'))` to `new MarkdownIt`.

[the remark this answers is not in the document]
Comment (yzhang-gh) [reply src/markdownEngine.ts:6]: Let's do it and :shipit:

**Comment by yzhang-gh** [comment]
Just merge this in when you are ready 😉

**Review by Lemmingh — APPROVED** [review]

**Comment by yzhang-gh** [comment]
Help yourself 👌

**Comment by yzhang-gh** [comment]
🍻

**Comment by Lemmingh** [comment]
😂

BTW, could you please disable the **merge commits** option?
**Settings** > **Options** > **Merge button**, uncheck **Allow merge commits**.

Such a long drop down menu is a visual clutter to me.

**Comment by yzhang-gh** [comment]
Ah, didn't know about this option.

**Pull request merged**